### PR TITLE
fix: Keep media update timeout alive so live playlists can recover

### PR DESCRIPTION
**Pull request by brandonocasey**
## Description
Right now for live hls streams we can be stuck buffering even though we should have recovered because we don't attempt to retry requests. You can reproduce this:

1. Select `Unified Streaming Live HLS` 
2. Play
3. Set network to offline in dev tools
4. Wait for the player to stall
5. Set network back to normal in dev tools
6. Player will never recover

This happens because we exclude and pause the current playlist loader when it fails a request.  Then we switch to same playlist again and since it is the same playlist we don't try to request it again. So now we have no mediaUpdateTimeout and we did not request the playlist again.

## The fix
1. We need a way to make sure that we start a mediaUpdateTimeout in `media`
2. We need mediaUpdateTimeouts to create media update timeouts 
3. We do not need to **always** return `true` in `shouldSwitchToMedia` for live playlists if the playlist is the same. This gave us state we have now as we were relying on abr to refresh live playlists in some scenarios.

### CODE REVIEW

File: src/playlist-loader.js [hunk 1]
```diff
@@ -705,14 +705,20 @@ export default class PlaylistLoader extends EventTarget {
...
+      window.clearTimeout(this.mediaUpdateTimeout);
+      this.mediaUpdateTimeout = null;
+    }
     const media = this.media();
 
     if (shouldDelay) {
       const delay = media ? ((media.partTargetDuration || media.targetDuration) / 2) * 1000 : 5 * 1000;
 
-      this.mediaUpdateTimeout = window.setTimeout(() => this.load(), delay);
+      this.mediaUpdateTimeout = window.setTimeout(() => {
```
Comment (brandonocasey) [inline src/playlist-loader.js:717]: imo this should be called something else, but I think we should save it for the playlist-loader base class refactor as dash-playlist-loader does the same thing.

**Review by gkatsev — APPROVED** [review]
Haven't tested yet but changes make sense to me.

File: src/master-playlist-controller.js [hunk 2]
```diff
@@ -67,7 +67,7 @@ const shouldSwitchToMedia = function({
   // This is because in LIVE, the player plays 3 segments from the end of the
   // playlist, and if `BUFFER_LOW_WATER_LINE` is greater than the duration availble
   // in those segments, a viewer will never experience a rendition upswitch.
-  if (!currentPlaylist || !currentPlaylist.endList) {
+  if (!currentPlaylist) {
     log(`${sharedLogLine} as current playlist ` + (!currentPlaylist ? 'is not set' : 'is live'));
```
Comment (gesinger) [inline src/master-playlist-controller.js:71]: Per the comment above, will this mean that we should have a way of ignoring the low water for live in the rest of this function? 

And we may need to update this line and the comment above

File: src/playlist-loader.js [hunk 3]
```diff
@@ -619,6 +613,8 @@ export default class PlaylistLoader extends EventTarget {
       return;
     }
 
+    this.updateMediaUpdateTimeout_(refreshDelay(playlist, true));
```
Comment (gesinger) [inline src/playlist-loader.js:616]: May be worth a comment above here saying why we update the timeout even if it's switching to the active playlist

File: test/master-playlist-controller.test.js [hunk 4]
```diff
@@ -2033,36 +2033,6 @@ QUnit.test(
...
-    this.masterPlaylistController.mainSegmentLoader_.trigger('bandwidthupdate');
-    assert.equal(
-      mediaChanges.length,
-      3,
-      'changes live media when sufficient forward buffer and higher ' +
-               'bandwidth playlist'
-    );
-
-    mediaChanges.length = 0;
-
```
Comment (gesinger) [inline test/master-playlist-controller.test.js:2065]: Do we want to remove these tests?

Comment (brandonocasey) [reply test/master-playlist-controller.test.js:2065]: These only tested that live content always switches as far as I could tell. We might have to add one more back in depending on changes that I have to make in shouldSwitchToMedia.

File: src/master-playlist-controller.js [hunk 5]
```diff
@@ -351,7 +356,7 @@ export class MasterPlaylistController extends videojs.EventTarget {
   checkABR_() {
     const nextPlaylist = this.selectPlaylist();
 
-    if (this.shouldSwitchToMedia_(nextPlaylist)) {
+    if (nextPlaylist && this.shouldSwitchToMedia_(nextPlaylist)) {
```
Comment (brandonocasey) [inline src/master-playlist-controller.js:359]: minor fix here. I saw that we throw an error if `nextPlaylist` is undefined.

File: src/playlist-loader.js (hunk 3 above)
Comment (brandonocasey) [reply src/playlist-loader.js:621]: added

File: src/playlist-loader.js [hunk 6]
```diff
@@ -705,14 +710,20 @@ export default class PlaylistLoader extends EventTarget {
...
+    }
     const media = this.media();
 
     if (shouldDelay) {
       const delay = media ? ((media.partTargetDuration || media.targetDuration) / 2) * 1000 : 5 * 1000;
 
-      this.mediaUpdateTimeout = window.setTimeout(() => this.load(), delay);
+      this.mediaUpdateTimeout = window.setTimeout(() => {
+        this.mediaUpdateTimeout = null;
+        this.load();
```
Comment (gkatsev) [inline src/playlist-loader.js:724]: is it hard to add a test for these lines?

**Review by gkatsev — COMMENTED** [review]
Can confirm that this fixes the issue of going offline and coming back!

File: src/playlist-loader.js (hunk 6 above)
Comment (brandonocasey) [reply src/playlist-loader.js:724]: i'll see what I can do.

Comment (brandonocasey) [reply src/playlist-loader.js:724]: added

**Review by gkatsev — APPROVED** [review]

**Review by gesinger — APPROVED** [review]

**Pull request merged**